Docs for user defined env vars validation (#2450)
wasp-lang/wasp: web/docs/project/env-vars.md

## Custom Env Var Validations

If your code requires some environment variables, you usually want to ensure that they are correctly defined. In Wasp, you can define your environment variables validation by defining a [Zod object schema](https://zod.dev/?id=basic-usage) and telling Wasp to use it.

:::info What is Zod?

> [Zod](https://zod.dev/) is a library that lets you define what you expect from your data. For example, you can use Zod to define that a value should be a string that's a valid email address or that a value should be a number between 0 and 100.

Reviewer: ```suggestion
[Zod](https://zod.dev/) is a library that lets you define what you expect from your data. For example, you can use Zod to define that:
- A value should be a string that's a valid email address.
- A value should be a number between 0 and 100.
- etc.
```
Plus, maybe add an explanation why that is helpful (something along the lines of "this is great because it warns us about improperly defined environment variables as soon as possible instead of failing somewhere else down the line").

:::

> Let's look at an example of defining env vars validation:

Reviewer: ```suggestion
Take a look at an example of defining env vars validation:
```
Addresses the user directly, but no big deal. We use "let's take a look at" all the time, although we might consider switching it up for something like this.

<Tabs groupId="js-ts">
<TabItem value="js" label="JavaScript">

```js title="src/env.js"
import * as z from 'zod'

export const serverEnvValidationSchema = z.object({
  STRIPE_API_KEY: z.string({
    required_error: 'STRIPE_API_KEY is required.',
  }),
})

export const clientEnvValidationSchema = z.object({
  REACT_APP_NAME: z.string().default('TODO App'),

> })

Reviewer: The API should not change depending on whether the user uses TS or JS. We should always tell users to call `defineEnvValidationSchema`.

Since JS and TS don't differ in runtime, neither should our API. The only differences should be in the type space.

It's better for both us and our users:
- Users will appreciate (or maybe even expect) the consistency.
- Two APIs mean maintaining two runtime behaviors/execution paths. The difference might be trivial, but it's there. Plus, we might one day decide to do something extra in `defineEnvValidationSchema`.

I noticed the same pattern with `userSignupFields`, and I'd like to change it there too (and all other places, if there are more).

> [!NOTE]
> I know [Vue does it like this](https://vuejs.org/guide/typescript/overview#definecomponent), but I still don't like it :)
>
> I guess they did it to stay somewhat backward-compatible. If they could start over, they'd probably make it consistent (but this is just my theory, maybe they prefer not to burden the JS users).

Author: Good point 👍 

I agree with both this and the `userSignupFields` change, created an issue: https://github.com/wasp-lang/wasp/issues/2455

```

</TabItem>
<TabItem value="ts" label="TypeScript">

```ts title="src/env.ts"
import * as z from 'zod'

import { defineEnvValidationSchema } from 'wasp/env'

export const serverEnvValidationSchema = defineEnvValidationSchema(
  z.object({
    STRIPE_API_KEY: z.string({
      required_error: 'STRIPE_API_KEY is required.',
    }),
  })
)

export const clientEnvValidationSchema = defineEnvValidationSchema(
  z.object({
    REACT_APP_NAME: z.string().default('TODO App'),
  })
)
```

> We are using the `defineEnvValidationSchema` helper to get type-checking for the Zod schema.

Reviewer: I suggest changing the subject to emphasize a point and address the user directly:
```suggestion
The defineEnvValidationSchema` function ensures your Zod schema is type-checked.
```

</TabItem>
</Tabs>

```wasp title="main.wasp"
app myApp {
  ...
  client: {
    envValidationSchema: import { clientEnvValidationSchema } from "@src/env",
  },
  server: {
    envValidationSchema: import { serverEnvValidationSchema } from "@src/env",
  },
}
```

> We defined schemas for both the client and the server env vars and told Wasp to use them. Wasp merges your env validation schemas with the built-in env vars validation schemas when it validates the `process.env` object on the server and the `import.meta.env` object on the client.

Reviewer: First sentence: we -> you.

The rest of the text already uses the second person, so that's great.

This means you can use the `env` object to access **your env vars** like this:

```ts title="src/stripe.ts"
import { env } from 'wasp/server'

const stripeApiKey: string = env.STRIPE_API_KEY
```

Read more about the env object in the [API Reference](#api-reference).

## API Reference

> There are **Wasp-defined** and **user-defined** env vars. Wasp has built-in validation for its env vars and you can define your own validation for your env vars.

Reviewer: I'd flip the ordering to better emphasize the main point:
```suggestion
There are **Wasp-defined** and **user-defined** env vars. Wasp already comes with built-in validation for Wasp-defined env vars. For your env vars, you can define your own validation.
```

```ts title="src/App.ts"
import { env } from 'wasp/client'

// Wasp-defined
const apiUrl: string = env.REACT_APP_API_URL

// User-defined
const analyticsId: string = env.REACT_APP_ANALYTICS_ID
```

</TabItem>
</Tabs>

> You can use `import.meta.env.REACT_APP_SOME_VAR_NAME` directly in your code, but it's not recommended because it's not validated and can lead to runtime errors if the env var is not defined.

Reviewer: ```suggestion
You can use `import.meta.env.REACT_APP_SOME_VAR_NAME` directly in your code. We don't recommend this. `import.meta.env` isn't validated and missing env vars can cause runtime errors.
```

import * as z from 'zod'

import { defineEnvValidationSchema } from 'wasp/env'

export const envValidationSchema = defineEnvValidationSchema(
  z.object({
    STRIPE_API_KEY: z.string({
      required_error: 'STRIPE_API_KEY is required.',
    }),
  })
)
```

We are using the `defineEnvValidationSchema` helper to get type-checking for the Zod schema.

</TabItem>
</Tabs>

```wasp title="main.wasp"
app myApp {
  ...
  server: {
    envValidationSchema: import { envValidationSchema } from "@src/env",
  },
}
```

Wasp merges your env validation schemas with the built-in env vars validation schemas when it validates the `process.env` object.

#### Accessing env vars in server code

You can access both **Wasp-defined** and **user-defined** client env vars in your client code using the `env` object:

<Tabs groupId="js-ts">
<TabItem value="js" label="JavaScript">

```js title="src/stripe.js"
import { env } from 'wasp/server'

// Wasp-defined
const serverUrl = env.WASP_SERVER_URL

// User-defined
const stripeApiKey = env.STRIPE_API_KEY
```

</TabItem>
<TabItem value="ts" label="TypeScript">

```ts title="src/stripe.ts"
import { env } from 'wasp/server'

// Wasp-defined

> const serverUrl: string = env.WASP_SERVER_URL

Reviewer: Why the `: string` part?

Maybe you wanted to emphasize that "this is now TypeScript." I understand the motivation, but we should document the TS code our users should write. If that's the same as JS code, so be it.

It's even better if the type-checker does its job silently. The fewer changes happening when you toggle JS/TS in the docs, the better - it means TS and Wasp are doing a good job.

Applies to other code boxes too.

Author: I had the intent of showing it's no longer `string | undefined` but actually a validated `string` but if that didn't come across, I'll remove it 👍

// User-defined
const stripeApiKey: string = env.STRIPE_API_KEY
```

</TabItem>
</Tabs>

> You can use `process.env.SOME_SECRET` directly in your code, but it's not recommended because it's not validated and can lead to runtime errors if the env var is not defined.

Reviewer: Same suggestion as above. This sentence is kind of a mouthful (that's what she said).
